### DISCUSSION TITLE
Custom loss function for multioutput regression

### DISCUSSION THREAD

**Discussion by min858** [dmlc/xgboost#11379] [Q&A]
Hello!

I am very new to XGBoost, so please excuse me if I ignore some obvious piece of knowledge.

I am currently working on a problem with a longitudinal setting where I am trying to predict multiple outputs.

In longitudinal data, off-diagonal terms are essential for reflecting intra-subject correlation. However, I noticed that we should use a diagonal approximation for hessian in the custom loss. 

I want to use deviance loss for my custom loss using a covariance matrix for each subject. Is there a way to use full hessian in the multioutput regression in XGBoost?

Thank you for your answer.

**Comment by trivialfis** [comment]
Hi, I'm not familiar with longitudinal data analysis, any suggested reading?

> However, I noticed that we should use a diagonal approximation for hessian in the custom loss.

Yes, otherwise, the Hessian becomes a 3-dim cube for the dataset, assuming targets are strongly associated. XGBoost currently handles only vector and matrix gradients.

> off-diagonal terms are essential for reflecting intra-subject correlation

The off-diagonal of the Hessian matrix?

> I want to use deviance loss for my custom loss using a covariance matrix for each subject.

Could you please share the math (equations)?

**Reply by min858** [reply]
Sorry to reply you late.

For example, my deviance loss is given by:

![deviance loss](https://latex.codecogs.com/png.image?\dpi{130}&space;\sum_{i}\left(\log\det(\Sigma_i)&plus;(y_i-\hat{y}_i)^T\Sigma_i^{-1}(y_i-\hat{y}_i)\right))

Here, ![y_i](https://latex.codecogs.com/png.image?\dpi{130}&space;y_i) is an ![m \times 1](https://latex.codecogs.com/png.image?\dpi{130}&space;m%20%5Ctimes%201) vector containing the repeated measurements for subject ![i](https://latex.codecogs.com/png.image?\dpi{130}&space;i), and ![\Sigma_i](https://latex.codecogs.com/png.image?\dpi{130}&space;%5CSigma_i) is the corresponding ![m \times m](https://latex.codecogs.com/png.image?\dpi{130}&space;m%20%5Ctimes%20m) covariance matrix for subject ![i](https://latex.codecogs.com/png.image?\dpi{130}&space;i).

Then, according to my calculation, the gradient with respect to ![\hat{y}_i](https://latex.codecogs.com/png.image?\dpi{130}&space;\hat{y}_i) is:

![gradient](https://latex.codecogs.com/png.image?\dpi{130}&space;\nabla_{\hat{y}_i}F=2\Sigma_i^{-1}(\hat{y}_i-y_i))

and the Hessian becomes:

![hessian](https://latex.codecogs.com/png.image?\dpi{130}&space;\nabla^2_{\hat{y}_i}F=\mathrm{blockdiag}(2\Sigma_1^{-1},\dots,2\Sigma_n^{-1}))

where ![n](https://latex.codecogs.com/png.image?\dpi{130}&space;n) is the number of subjects.

When computing the loss (e.g., for splitting), I want it to be evaluated at the **subject level**, as above. (This is because repeated measurements in the same subject depend on each other.)

However, I want the prediction itself to be **univariate**, that is, to predict ![y_{ij}](https://latex.codecogs.com/png.image?\dpi{130}&space;y_{ij}), the *j*-th repeated measurement of the *i*-th subject.

Can such a loss function be defined in standard univariate XGBoost?

I originally intended to perform multi-output regression, but due to a change in the direction of the study, I'm now asking this question instead.

Thank you in advance!

**Reply by trivialfis** [reply]
It might take some time for me to go through your gradient calculation.

> This is because repeated measurements in the same subject depend on each other

Each measurement depends on other measurements on the same subject? I don't think that's possible? Shouldn't repeated measurements be independent? Or is the dependency about different subjects?

> I want the prediction itself to be univariate,

I think it's possible by using one model for each target. The gradient is calculated together for all targets, then trees are grown independently.
